zcash_client_backend/src/welding_rig.rs

```diff
@@ -68,38 +66,111 @@ fn scan_output<P: consensus::Parameters>(
         // - Notes sent from one account to itself.
         let is_change = spent_from_accounts.contains(&account);
 
+        let witness = IncrementalWitness::from_tree(tree);
+        let nf = vk.nf(&note, &witness);
+
         return Some(WalletShieldedOutput {
             index,
             cmu,
             epk,
-            account: *account,
+            account: **account,
             note,
             to,
             is_change,
-            witness: IncrementalWitness::from_tree(tree),
+            witness,
+            nf,
         });
     }
+
     None
 }
 
-/// Scans a [`CompactBlock`] with a set of [`ExtendedFullViewingKey`]s.
+/// A key that can be used to perform trial decryption and nullifier
+/// computation for a Sapling [`CompactOutput`]
+///
+/// [`CompactOutput`]: crate::proto::compact_formats::CompactOutput
+pub trait ScanningKey {
+    type Nf;
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)>;
+
+    fn nf(&self, note: &Note, witness: &IncrementalWitness<Node>) -> Self::Nf;
+}
+
+impl ScanningKey for ExtendedFullViewingKey {
+    type Nf = Nullifier;
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)> {
+        try_sapling_compact_note_decryption(params, height, &self.fvk.vk.ivk(), &epk, &cmu, &ct)
+    }
+
+    fn nf(&self, note: &Note, witness: &IncrementalWitness<Node>) -> Self::Nf {
+        note.nf(&self.fvk.vk, witness.position() as u64)
+    }
+}
+
+impl ScanningKey for SaplingIvk {
+    type Nf = ();
+
+    fn try_decryption<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        height: BlockHeight,
+        epk: &jubjub::ExtendedPoint,
+        cmu: &bls12_381::Scalar,
+        ct: &[u8],
+    ) -> Option<(Note, PaymentAddress)> {
+        try_sapling_compact_note_decryption(params, height, self, &epk, &cmu, &ct)
+    }
+
+    fn nf(&self, _note: &Note, _witness: &IncrementalWitness<Node>) {}
```

Comment:
(no change required, just a comment)
minor observation: given that this is a no-op is `nf` really a function of the ScanningKey trait or is it unique to ExtFvks? Since the nullifier implementation in extfvk relies on the note and witness, it seems like it might be possible to exclude the `nf` function from this trait.

Reply:
The reason that this has to be a member of the trait is that the whole purpose of the trait is to allow us to specialize the type of the result of scan_block to reflect whether or not a nullifier was computed. This depends upon the type of scanning key that was passed to the function. So here we have an example where the *result type* of the function depends upon the type of the argument to the function. This is really desirable, because you don't have to represent the presence or absence of the nullifier as an `Option` - instead, it's statically known to be either present or absent.